### PR TITLE
Add new aggregation pipeline stages

### LINKED ISSUE #1618: Aggregation builder: Add support for aggregation pipeline stages added in MongoDB 3.4
See doctrine/mongodb#288, doctrine/mongodb#292 and doctrine/mongodb#293.

Main work required affects supporting `$graphLookup` in combination with simple references.

Also, merging #1566 would also allows us to use non-DBRef-references in `$lookup` and `$graphLookup` as well.

### PR DESCRIPTION
**Pull request by alcaeus**
Fixes #1618.

This pull request adds support for the new aggregation pipeline stages added in MongoDB 3.4 by supporting them with mapping information. It also fixes field name translation in `$lookup` stages.

Todos:
- [x] Add documentation for new pipeline stages including examples
- [x] Update `$lookup` limitations about reference types; document support for `storeAs=ref`

### CODE REVIEW

File: lib/Doctrine/ODM/MongoDB/Aggregation/Stage/Bucket.php [hunk 1]
```diff
@@ -0,0 +1,49 @@
+<?php
+
```
Comment (malarzm) [inline lib/Doctrine/ODM/MongoDB/Aggregation/Stage/Bucket.php]: Missing license header

File: lib/Doctrine/ODM/MongoDB/Aggregation/Stage/BucketAuto.php [hunk 2]
```diff
@@ -0,0 +1,49 @@
+<?php
+
```
Comment (malarzm) [inline lib/Doctrine/ODM/MongoDB/Aggregation/Stage/BucketAuto.php]: Missing license header

File: lib/Doctrine/ODM/MongoDB/Aggregation/Stage/GraphLookup.php [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+            throw MappingException::cannotUseShardedCollectionInLookupStages($this->targetClass->name);
+        }
+
+        return parent::from($this->targetClass->getCollection());
+    }
+
+    public function connectFromField($connectFromField)
+    {
+        // No targetClass mapping - simply use field name as is
+        if (!$this->targetClass) {
```
Comment (malarzm) [inline lib/Doctrine/ODM/MongoDB/Aggregation/Stage/GraphLookup.php]: Missing spaces around `!`

File: lib/Doctrine/ODM/MongoDB/Aggregation/Stage/GraphLookup.php [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+            return parent::connectFromField($connectFromField);
+        }
+
+        // connectFromField doesn't have to be a reference - in this case, just convert the field name
+        if (!$this->targetClass->hasReference($connectFromField)) {
```
Comment (malarzm) [inline lib/Doctrine/ODM/MongoDB/Aggregation/Stage/GraphLookup.php]: Missing spaces around `!`

File: lib/Doctrine/ODM/MongoDB/Aggregation/Stage/GraphLookup.php [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+            }
+
+            parent::connectFromField($referencedFieldName);
+        } else {
+            if (isset($referenceMapping['repositoryMethod'])) {
+                throw MappingException::repositoryMethodLookupNotAllowed($this->class->name, $connectFromField);
+            }
+
+            $mappedByMapping = $this->targetClass->getFieldMapping($referenceMapping['mappedBy']);
+            switch ($mappedByMapping['storeAs']) {
```
Comment (malarzm) [inline lib/Doctrine/ODM/MongoDB/Aggregation/Stage/GraphLookup.php]: These two `switch`es are almost the same, what do you think about rewriting whole `if-else` block to:

```
$mapping = $referenceMapping['isOwningSide'] ? $referenceMapping : $this->targetClass->getFieldMapping($referenceMapping['mappedBy']);
// if for repository method
// whole switch
```

File: lib/Doctrine/ODM/MongoDB/Aggregation/Stage/GraphLookup.php [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+                    throw MappingException::cannotLookupDbRefReference($this->class->name, $connectFromField);
+            }
+
+            parent::connectFromField($referencedFieldName);
+        } else {
+            if (isset($referenceMapping['repositoryMethod'])) {
+                throw MappingException::repositoryMethodLookupNotAllowed($this->class->name, $connectFromField);
+            }
+
+            $mappedByMapping = $this->targetClass->getFieldMapping($referenceMapping['mappedBy']);
```
Comment (malarzm) [inline lib/Doctrine/ODM/MongoDB/Aggregation/Stage/GraphLookup.php]: Should there be an exception for when `mappedBy` is not present? Or is that `mappedBy` or `repositoryMethod`?

File: lib/Doctrine/ODM/MongoDB/Aggregation/Stage/GraphLookup.php [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+        $referenceMapping = $this->class->getFieldMapping($fieldName);
+        $this->targetClass = $this->dm->getClassMetadata($referenceMapping['targetDocument']);
+        if ($this->targetClass->isSharded()) {
+            throw MappingException::cannotUseShardedCollectionInLookupStages($this->targetClass->name);
+        }
+
+        parent::from($this->targetClass->getCollection());
+
+        if ($referenceMapping['isOwningSide']) {
+            switch ($referenceMapping['storeAs']) {
```
Comment (malarzm) [inline lib/Doctrine/ODM/MongoDB/Aggregation/Stage/GraphLookup.php]: Same switch again, maybe a private method accepting `array $mapping` would be better?

File: lib/Doctrine/ODM/MongoDB/Aggregation/Stage/GraphLookup.php [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+        }
+    }
+
+    protected function convertTargetFieldName($fieldName)
+    {
+        if (is_array($fieldName)) {
+            return array_map([$this, 'convertTargetFieldName'], $fieldName);
+        }
+
+        if (!$this->targetClass) {
```
Comment (malarzm) [inline lib/Doctrine/ODM/MongoDB/Aggregation/Stage/GraphLookup.php]: Missing spaces around `!`

File: lib/Doctrine/ODM/MongoDB/Mapping/MappingException.php [hunk 9]
```diff
@@ -396,4 +405,9 @@ public static function referencePrimersOnlySupportedForInverseReferenceMany($cla
     {
         return new self("Cannot use reference priming on '$fieldName' in class '$className'. Reference priming is only supported for inverse references");
     }
+
+    public static function connectFromFieldMustReferenceSameDocument($fieldName)
+    {
+        return new self("Cannot use field '$fieldName' as connectFromField in \$graphLookup stage. Reference must target the document itself.");
```
Comment (malarzm) [inline lib/Doctrine/ODM/MongoDB/Mapping/MappingException.php]: if I'm not mistaken, you missed an "a": "in *a* $graphLookup stage"

File: lib/Doctrine/ODM/MongoDB/Aggregation/Stage/GraphLookup.php (hunk 6 above)
Comment (alcaeus) [reply lib/Doctrine/ODM/MongoDB/Aggregation/Stage/GraphLookup.php]: There's an exception if `repositoryMethod` is present - I don't specifically check for `mappedBy` here. It'd probably be good from a defensive coding point of view. 👍

File: lib/Doctrine/ODM/MongoDB/Aggregation/Stage/GraphLookup.php (hunk 7 above)
Comment (alcaeus) [reply lib/Doctrine/ODM/MongoDB/Aggregation/Stage/GraphLookup.php]: Done - not sure if I want to extract it into a trait for use in Stage\Lookup as well :|

**Pull request merged**